core/src/main/java/org/ldaptive/control/util/SyncReplRunner.java

```diff
@@ -94,44 +65,136 @@ public void accept(final Exception e)
   /** Invoked when a sync info message is received. */
   private Consumer<SyncInfoMessage> onMessage;
 
+  /** Invoked when an exception occurs. */
+  private Consumer<Exception> onException;
+
   /** Whether the sync repl search is running. */
   private boolean started;
 
-  /** Prevent multiple invocations of onException. */
-  private AtomicBoolean handlingException = new AtomicBoolean();
-
 
   /**
-   * Creates a new sync repl runner. Uses a custom {@link ConnectionFactoryTransport} for processing I/O and messages.
+   * Creates a new sync repl runner. The supplied connection factory is modified to invoke {@link
+   * SyncReplClient#send(SearchRequest, CookieManager)} when the connection opens and {@link SyncReplClient#cancel()}
+   * when the connection closes.
    *
-   * @param  config  sync repl connection configuration
+   * @param  cf  to get a connection from
    * @param  request  sync repl search request
    * @param  manager  sync repl cookie manager
    */
-  public SyncReplRunner(final ConnectionConfig config, final SearchRequest request, final CookieManager manager)
+  public SyncReplRunner(final SingleConnectionFactory cf, final SearchRequest request, final CookieManager manager)
+  {
+    syncReplClient = new SyncReplClient(cf, true);
+    searchRequest = request;
+    cookieManager = manager;
+    cf.setOnOpen(conn -> {
+      try {
+        syncReplClient.send(searchRequest, cookieManager);
+      } catch (LdapException e) {
+        LOGGER.error("Could not send sync repl request", e);
+        return false;
+      }
+      return true;
+    });
+    cf.setOnClose(conn -> {
+      try {
+        if (!syncReplClient.isComplete()) {
+          syncReplClient.cancel();
+        }
+      } catch (Exception e) {
+        LOGGER.warn("Could not cancel sync repl request", e);
+        return false;
+      }
+      return true;
+    });
+  }
+
+
+  /**
+   * Creates a new single connection factory. Uses a {@link SearchConnectionValidator} for connection validation. See
+   * {@link #createTransport()}
+   *
+   * @param  config  sync repl connection configuration
+   *
+   * @return  single connection factory for use with a sync repl runner
+   */
+  public static SingleConnectionFactory createConnectionFactory(final ConnectionConfig config)
```

Comment:
While I can imagine these are helpful, their presence in this class strikes me as odd. Consider a separate class with convenience constructors; compare with the `Executors` JDK class.

Reply:
Would a new class called `ConnectionFactories` make more sense? With methods like:
- `ConnectionFactories#newSyncReplRunnerFactory(ConnectionConfig)`
- `ConnectionFactories#newSyncReplRunnerFactory(ConnectionConfig, ConnectionValidator)`
- `ConnectionFactories#newSyncReplRunnerFactory(Transport, ConnectionConfig, ConnectionValidator)`

Reply:
Yes, that's in line with my thinking.

Reply:
Thinking on this some more, I'm not inclined to add a separate helper class _yet_. You already noted that we only have a single use case. I'd prefer for one or two more use cases to arrive before moving those methods out of the `SyncReplRunner`. They arguably make more sense where they are today.